### PR TITLE
dns: add dns.rcode keyword v2

### PR DESCRIPTION
**Pull request by hadiqaalamdar**
Feature #6621

- [x] I have read the contributing guide lines at https://docs.suricata.io/en/latest/devguide/codebase/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/issues/6621) ticket: https://redmine.openinfosecfoundation.org/issues/6621

Previous PR: https://github.com/OISF/suricata/pull/10087

Describe changes:
- added the dns.rcode field
- made the suggested changes from previous PR
- successfully ran the `cargo clippy --all-features` command as recommended.

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1567

### CODE REVIEW

**Review by inashivb — COMMENTED** [review]
Good work, Hadiqa!
Are you able to see s-v failures locally too?

**Comment by hadiqaalamdar** [comment]
> Good work, Hadiqa! Are you able to see s-v failures locally too?

yes, so the SV tests for rcode seem to all be failing.

**Comment by inashivb** [comment]
> yes, so the SV tests for rcode seem to all be failing.

I see. Do you need help with that? Could you please share if you have an idea why it might be happening?

File: src/detect-dns-rcode.c [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        void *txv, const Signature *s, const SigMatchCtx *ctx)
+{
+    return rs_dns_rcode_match(txv, (void *)ctx, flags);
+}
+
+void DetectDnsRcodeRegister(void)
+{
+    sigmatch_table[DETECT_AL_DNS_RCODE].name = "dns.rcode";
+    sigmatch_table[DETECT_AL_DNS_RCODE].desc = "Match the DNS header rcode flag.";
+    sigmatch_table[DETECT_AL_DNS_RCODE].url = "/rules/dns-keywords.html#dns-rcode";
```
Comment (jufajardini) [inline src/detect-dns-rcode.c:74]: Nit: once the code-related changes are done, we'll need to populate the documentation section mentioned in this line ;)

File: src/detect-dns-rcode.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/* Copyright (C) 2019 Open Information Security Foundation
```
Comment (jufajardini) [inline src/detect-dns-rcode.h:1]: Nit: year 2024

File: src/detect-dns-rcode.c [hunk 3]
```diff
@@ -0,0 +1,87 @@
+/* Copyright (C) 2023 Open Information Security Foundation
```
Comment (jufajardini) [inline src/detect-dns-rcode.c:1]: Nit year 2024, now :P

**Review by jufajardini — CHANGES_REQUESTED** [review]
As Shivani has indicated, could you go over the CI failures and share with us if you have an idea of why are the tests failing?

I've also left a review on the SV PR that may add insights on this.

For this PR, for now, just some nit comments inline.

**Comment by hadiqaalamdar** [comment]
> As Shivani has indicated, could you go over the CI failures and share with us if you have an idea of why are the tests failing?
> 
> I've also left a review on the SV PR that may add insights on this.
> 
> For this PR, for now, just some nit comments inline.

I don't really know why the failures are happening. I didn't see any errors when I ran the code locally. I'll create a new PR and see if the continue to happen.

**Comment by jufajardini** [comment]
> > As Shivani has indicated, could you go over the CI failures and share with us if you have an idea of why are the tests failing?
> > I've also left a review on the SV PR that may add insights on this.
> > For this PR, for now, just some nit comments inline.
> 
> I don't really know why the failures are happening. I didn't see any errors when I ran the code locally. I'll create a new PR and see if the continue to happen.

Can you run the suricata verify tests locally and see what is the output in the eve.json logs? If the checks are failing, comparing what is on the checks with what the logs actually looks like is one of the first things I do, if there are no errors, to debug a failing test.

**Comment by hadiqaalamdar** [comment]
New PR: https://github.com/OISF/suricata/pull/10126

**Pull request closed without merging**